### ISSUE TITLE
NSL rate (sync marker)

### ISSUE THREAD

**Issue by egaltier** [pcdshub/mec#20]
**Feature Request**
To accomodate a new operation script that take advantage of the VISAR that can be saved in the DAQ, we would need to run the sync marker at 1Hz before calling `x.nsl.shot`.

**Existing Alternatives**
For now there is none since the value is protected and set to 10 Hz. At the moment and at this rate, it will take one image, and the rest will be damaged. Not a show stopper but it brings instability to the DAQ.

**Context**
In the script we use to take VISAR references, the rate in the `x.nsl.shot`  plan is set to 10 Hz, to match the one needed for the long pule laser. However, the VISAR cannot run this fast. 1 Hz is kind of the optimum rate. Since we are combining the X-ray only with the visar reference, I would like to set the rate at 1 Hz so that the VISAR doesn't get damaged events when taking say, 10 events (which would produce 10 visar references). Technically, I should just update the dictionary of the config function like `x.nsl.rate=1` or `x.nsl.config['rate']=1` , but it seems that my call is overwritten somewhere else. I think it is overwritten because the long pulse laser needs the sync mark at 10 Hz to run properly. But we don't use it in this script. 

Just to be clear: the long pulse laser NEEDS to run on the 10 Hz sync marker, but when we take references we only use the VISAR and the X-ray laser so this restriction is not necessary, and we would prefer to run on the 1Hz sync marker to prevent VISAR damage event.

If you lift the restriction completely, we need to know that to make sure we force 10 Hz before we take a driven shot (not a big deal, just one line of code).

**Comment by egaltier** [comment]
Discussing with Nick, it turns out I just had to use `x.nsl._config['rate'] = 1` to set the rate after initialization. So I tested the scripts and they do behave very closely to what I want (the DAQ doesn't make damaged event anymore and we save the right number of images). There is only one small behavior I would like to sort out. When I change the rate from 10 to 1 Hz, I want to make sure it is reversed back to 10 Hz just before the completion of the script which execute the rate change. So, I use `op.x.nsl._config['rate'] = 10` at the end of the script to make sure the configuration of the plan has the right rate (for whoever takes the beam after, if not the current operator). While it seems to work (`x.nsl.config` does return a rate of 10), the actual GUI shows that the sync marker is still at 1 Hz. I have the feeling that it is because I dont run a plan to push for the configuration.  If that's true, of course, I am not going to do this just to reset the rate back to 10 Hz. Would there be a less 'intrusive' way to make sure the sync marker is indeed back to 10 Hz?

**Comment by slactjohnson** [comment]
@egaltier if you change the configured rate back to 10, that will be pushed to the sequencer whenever the next shot plan is executed.

**Comment by egaltier** [comment]
Right, whenever the plan is executed. And I have check that this part works. I worry about this situation: we are done with a shift, but Eric C. wants to take some test shot for debugging purposes. He will not be using the scripts for the shots so there will be no plan that will push the sync marker back to 10 Hz. I have looked at his module and in theory it doesn't look like he is checking the sync marker at all. So I do believe we are ok. But it will be safer if we can make sure that the sync marker is pushed back to its original state, and trying to do so without pushing an entire plan (since I would assume this would be an overkill). Maybe we just toggle the value of the button's PV itself?

**Comment by slactjohnson** [comment]
It sounds like we need a special shot plan for this, as you suggested in your request. This should not be hard to do.
